processing/src/main/java/org/apache/druid/segment/filter/AndFilter.java

```diff
@@ -73,21 +73,119 @@ public AndFilter(List<Filter> filters)
     this(new LinkedHashSet<>(filters));
   }
 
-  public static <T> ImmutableBitmap getBitmapIndex(
-      ColumnIndexSelector selector,
+  @Override
+  public <T> FilterBundle makeFilterBundle(
+      ColumnIndexSelector columnIndexSelector,
       BitmapResultFactory<T> bitmapResultFactory,
-      List<Filter> filters
+      int selectionRowCount,
+      int totalRowCount,
+      boolean includeUnknown,
+      boolean allowPartialIndex
   )
   {
-    return bitmapResultFactory.toImmutableBitmap(
-        getBitmapIndex(selector, filters).computeBitmapResult(bitmapResultFactory, false)
+    final List<FilterBundle.IndexBundleInfo> indexBundles = new ArrayList<>();
+    final List<FilterBundle.MatcherBundle> matcherBundles = new ArrayList<>();
+
+    int selectionCount = selectionRowCount;
+    ImmutableBitmap index = null;
+
+    final long bitmapConstructionStartNs = System.nanoTime();
+    for (Filter subfilter : filters) {
+      final FilterBundle subBundle = subfilter.makeFilterBundle(
+          columnIndexSelector,
+          bitmapResultFactory,
+          selectionCount,
+          totalRowCount,
+          includeUnknown,
+          allowPartialIndex
+      );
+      if (subBundle.getIndex() != null) {
```

Comment:
In the code for generating the `Cursor` you were checking if the `FilterBundle` was null, here you are not.  Are you missing a check here or are you needlessly checking there?

Reply:
As mentioned above, `Filter.makeFilterBundle` cannot return null, it must return a bundle that has at least an index or a matcher. The cursor case is for when there is no filter on the cursor, but as mentioned in other comment will try to clean that up a bit